### ISSUE TITLE
ECS Physics

### ISSUE THREAD

**Issue by fire** [GodotECS/godex_physics#1]
---
name: Feature / Enhancement proposal
about: 'Open a new proposal'
title: ''
labels: ''
assignees: ''

---

### Describe the project you are working on
3d Game

### Describe the problem or limitation you are having in your project
Physics has to be faster

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
ECS can avoid the locking problem traditional physics.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

struct PhysicalProperties : public godex::Component {
    COMPONENT(DynamicalBody, DenseVector);

    uint32_t physical_id = UINT32_MAX;
    real_t mass = 1.0;
    real_t bounciness = 0.0;
    real_t friction = 0.5;
};

struct DynamicalBody : public godex::Component {
    COMPONENT(DynamicalBody, DenseVector);

    /* Just a tag */
};

struct Collider : public godex::Component {
    COMPONENT(DynamicalBody, DenseVector);
    
    /*shape_id or shape data.*/
};

How would we divide the world?

Would we have more than one?

How do we simulate the world?
```C++
void config_body(PhysicsWorld* p_physics_world, Query<Changed<PhysicalProperties>> p_query){
    while(p_query.is_done() == false){
        auto [phys] = p_query.get();
        p_physics_world.rigid_body[phys.physical_id].mass = phys.mass;
        p_physics_world.rigid_body[phys.physical_id].bounciness = phys.bounciness;
        p_physics_world.rigid_body[phys.physical_id].friction = phys.friction;
        p_query.next();
    }
}
```

The world is composed in this way:
You can have static body in the world by adding entities like this:
- Entity [PhysicalProperties, Collider]
- Entity [PhysicalProperties, Collider]
But you can also have rigid Body by adding the dynamical body tag component:
- Entity [PhysicalProperties, Collider, DynamicalBody]
- Entity [PhysicalProperties, Collider, DynamicalBody]
Basically you don't have to think about objects anymore. ECS is really another thing. Data and Functions are separate no more combined.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
Physics is performance sensitive and not a few lines of code.

### Is there a reason why this should be core and not an add-on in the asset library?
ECS is not in core, but it should be a C++ module.

**Comment by slapin** [comment]
if physics components can be quickly removed/added on the fly that would be good step forward, could be implemented using PhysicsServer for now.
